### PR TITLE
Use CLI to bundle packs for MRVA

### PR DESCRIPTION
**Pull request by dbartol**
This PR updates the VS Code extension to use `codeql pack bundle --mrva` to create the compiled pack to submit to MRVA. If the current CLI does not support `codeql pack bundle --mrva`, we fall back to the old manual pack creation path, where we manually copy the appropriate files and adjust the dependencies.

A commit-by-commit review is suggested. The first couple commits have the major changes. The remainder are either trivial fixes or minor test logging improvements. These test logging improvements were useful for debugging CI failures, so I left them in for future use.

Key changes:

#### Version testing
- The extension can now consume the JSON output of `codeql version` to see what features the current CLI version supports. We use this to detect `--mrva` support for `codeql pack bundle`. This is the preferred feature detection mechanism going forward.
- 
#### `generateQueryPack()`
`generateQueryPack()` for MRVA now has three main cases:
1. The selected query is in a pack, and the CLI supports `--mrva`. In this case, we just invoke the CLI to do the bundling.
2. The selected query is in a pack, and the CLI does not support `--mrva`. In this case, we do what we did before: copy the necessary parts of the pack to a temporary directory, try to fix up the dependencies as best we can, and then have the CLI bundle that temporary pack.
3. The selected query is not in a pack. Regardless of CLI support, we just create a temporary pack with that query and have the CLI bundle that temporary pack.

#### Test improvements
In the test code, I extended the Jest `expect` object to add `.toExistInPack()`. This tests whether a particular file path exists in the QL pack, and gives a better failure message if it does not.

#### 8.3 path expansion
Before we invoke `codeql pack bundle`, we now expand any 8.3 path components in the directory path on Windows. The `pack bundle` command ignores any files outside of the pack root, but current versions of the CLI wind up expanding 8.3 paths for the paths of the individual files, but do not expand the path of the pack root itself. This causes the CLI to think that, for example, `C:\SomeLongName\foo.ql` is not inside of `C:\SomeLo~1` and leave the file out of the bundled pack. 

I couldn't find any standard NodeJS file operation that will expand 8.3 paths, so I had to write one myself. My implementation traverses each 8.3 component in the path and enumerates the children of the parent directory, looking for one whose dev+inode matches the dev+inode for the short path.

By default, modern versions of Windows don't even generate 8.3 paths anymore. However:
1. The Actions runner images have opted in to 8.3 path generation.
2. The username under which the job runs on the worker is long (`runneradmin`).
3. The temp directory is under the `C:\Users\runneradmin` directory.
4. The `TEMP` environment variable uses the 8.3 (`C:\Users\runner~1`) version of the path.

This lead to failures that only reproduced in CI. Because local dev machines don't enable 8.3 path generation, I didn't add any specific tests for this. We're basically just testing it indirectly when we run the MRVA tests in CI.

### CODE REVIEW

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (charisk) [inline extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts:34]: Shall we move this to `/test/matchers`  (and maybe call it toExistInCodeQLPack`)?

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (charisk) [inline extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts:45]: Should we use `os.EOL` instead of `\n`? Similar comment for the message.

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/common/short-paths.ts:88]: I don't think 8.3 short names is a concept that will be familiar to many people. I hadn't heard of it, though thankfully once I'd recognised that it's a windows thing a google search did point me in the right direction.

Should we include more explanation or link to something like https://en.wikipedia.org/wiki/8.3_filename?

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:294]: How big of a problem do you expect this to be in practice? My research says they're a feature from pre-windows 95 and now only supported for compatibility with old systems. We don't consider these 8.3 filenames anywhere else, but should we? How likely are we to encounter them with users?

I think it's impressive that we're handling them and I don't want to discount the work you've done here, but the code in `short-paths.ts` is quite complex so I just want to be sure that the maintenance burden is going to be worth it compared to how many users it will unblock. Do you have any more insights or data on this?

File: extensions/ql-vscode/src/codeql-cli/cli.ts
Comment (charisk) [inline extensions/ql-vscode/src/codeql-cli/cli.ts:1423]: I don't see any usages of this. Is that intentional?

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/common/short-paths.ts:92]: Can we add some tests of this please? It's quite complex and esoteric, and since this code only runs on windows we're not going to be testing it in our day-to-day running (except for poor Shati who will hit all the bugs). So this would greatly benefit from some unit tests that run on windows.

File: extensions/ql-vscode/src/extension.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/extension.ts:443]: How confident are you you've found all of these places where we insert the version into a string? I haven't exhaustively gone through trying to find them, so I just thought I'd  ask how you feel.

If we miss any it'll lead to a classic `[Object object]` which wouldn't look too good.

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:82]:
```suggestion
    throw new UserCancellationException("Could not determine language");
```

I think the normal convention is to not use punctuation in error messages so they can more easily be composed together and stuff.

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:86]: Can we move this definition of `precompilationOpts` down to line 141 just before the `if-else` where it is defined? Might make it easier for future readers.

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:140]: Can we delete this `clearCache` call? We might have already just done it on line 136, or on line 103. So I think we'll have always already cleared after the last CLI command.

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:154]:
```suggestion
      const cache = join(originalPackRoot, ".cache");
```
Typo, or is this intentional?

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (robertbrignull) [inline extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts:40]:
```suggestion
    throw new TypeError(
      `Expected actual value to be a string. Found: ${typeof actual}`,
    );
```
So it'll be easier to debug if this fails.

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (robertbrignull) [inline extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts:44]: Technically we don't have to construct `files` or `filesString` except in the "fail" case. They could be moved into that branch.

File: extensions/ql-vscode/test/vscode-tests/cli.ts
Comment (robertbrignull) [inline extensions/ql-vscode/test/vscode-tests/cli.ts:11]:
```suggestion
      folder.uri.path.endsWith(`${os.sep}codeql`) || folder.uri.path.endsWith(`${os.sep}ql`),
```
I realise the original code is also using "/codeql", but since we're looking at it, should this be using `os.sep` or is forward slash always the right thing to use?

**Review by charisk — COMMENTED** [review]
This is generally looking good to me, although there's quite a lot going on so I could have easily missed something.

I have a few comments but the main functionality seems okay to me. I've also tested (with the currently released version of the CLI) and things seem to be working fine.

**Review by robertbrignull — COMMENTED** [review]
I've taken a look through. Generally looks good though I have quite a few comments because this is quite a large change to a few different areas.

I haven't gone through the short paths code in high details because it's an area I'm really not familiar with.

I've also done a manual test and it's correctly building a query pack and able to run MRVA, at least with CodeQL CLI 2.16.0. Have you done manual testing using old versions of the CodeQL CLI?

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (charisk) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:154]: This is the existing naming, I guess because it stands for compilation cache?

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (robertbrignull) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:294]: Sorry I didn't read the PR description well enough. I see you've explained that you hit this on the actions runners and why it's happening. In that case I see you're point that it's only really a case for CI and not for user machines. I still think some unit tests would be helpful, unless they're really hard to write. But I see the reason why this is implemented.

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (koesie10) [reply extensions/ql-vscode/src/common/short-paths.ts:7]: If we're not removing it, I'd suggest using the `BaseLogger` interface instead of using `extLogger` directly. All of the code in this file is independent of VS Code and could be tested using unit tests, but adding the `extLogger` import adds a dependency on the `vscode` module, which requires any tests to be located in the `vscode-tests` directory.

File: extensions/ql-vscode/test/vscode-tests/cli.ts
Comment (koesie10) [reply extensions/ql-vscode/test/vscode-tests/cli.ts:11]: I think the `uri` is normalized to always contain `/`. Only `fsPath` contains the platform-specific path separators.

However, could we simply use `dirname` here since that seems to be what we're checking against? Something like this (I kept it short to avoid introducing extra variables, but it might be a good idea to do explicit `===` checks instead of `includes`):

```suggestion
      ["codeql", "ql"].includes(dirname(folder.uri.fsPath)),
```

File: extensions/ql-vscode/src/codeql-cli/cli.ts
Comment (dbartol) [reply extensions/ql-vscode/src/codeql-cli/cli.ts:1423]: No, I think I must have added it before I realized we actually needed to call `pack bundle` instead. Removed.

Comment (dbartol) [reply extensions/ql-vscode/src/codeql-cli/cli.ts:1423]: Removed.

File: extensions/ql-vscode/src/codeql-cli/cli.ts
Comment (dbartol) [reply extensions/ql-vscode/src/codeql-cli/cli.ts:1437]: I believe there are some scenarios where `codeql pack bundle` can attempt to download packs from the registry, but I'm not sure which cases those are TBH. The CLI will only prompt for auth info if it actually does try to contact the registry, so this shouldn't affect the case where we're just accessing local packs.

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (dbartol) [reply extensions/ql-vscode/src/common/short-paths.ts:1]: Which repo should we track the tech debt in?
Any unit test we write would require manual machine configuration to run on a typical developer machine, since 8.3 name generation has been off by default for the last decade or so. It would still run in CI, though, since the Actions runners seem to have that feature enabled.

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (dbartol) [reply extensions/ql-vscode/src/common/short-paths.ts:6]: This code never gets executed on non-CI machines, so I found it very convenient to have one place where I could replace the `extlogger` call with `console.log` and have messages show up in the CI log directly. I can add a comment to this effect.

Is there a better way to see the extension log from a CI run? If so, I'll put that in a comment at the top of this file for anyone unlucky enough to have to debug it.

Comment (dbartol) [reply extensions/ql-vscode/src/common/short-paths.ts:6]: Nevermind, I've moved the code over to accept a `BaseLogger` object, so now we could just pass a console logger at the call site if we temporarily want console debugging. No need for the private `log()` function anymore.

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (dbartol) [reply extensions/ql-vscode/src/common/short-paths.ts:7]: Fixed.

Comment (dbartol) [reply extensions/ql-vscode/src/common/short-paths.ts:7]: This code is only executed on machines with a particular legacy feature enabled, and even then only if we actually see an 8.3 filename. Outside of our CI machines, which have this enabled, I expect that fewer than 1% of customers will ever get a single line logged from this code. On the other hand, if something does go wrong in this code, we're unlikely to notice without a log message, and it will be difficult to reproduce locally. So, I'd prefer to leave them in.

I did convert everything over to use `BaseLogger`, though, which solved one of Robert's comments and will make it easier to write a unit test in the future if we decide to.

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (dbartol) [reply extensions/ql-vscode/src/common/short-paths.ts:88]: Added a link.

Now I guess I'll crawl back in my cave and finish that painting of a mastodon I've been working on.

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (dbartol) [reply extensions/ql-vscode/src/common/short-paths.ts]: Good idea. Done.

Comment (dbartol) [reply extensions/ql-vscode/src/common/short-paths.ts]: See https://github.com/github/vscode-codeql/pull/3233/files#r1456142526

File: extensions/ql-vscode/src/extension.ts
Comment (dbartol) [reply extensions/ql-vscode/src/extension.ts:443]: I feel pretty confident. The only place the `VersionAndFeatures` gets exposed outside of `CodeQLCliServer` is via the version changed listener event (here). I made `CodeQLCliServer.getVersionAndFeatures` private, so anyone else will either call `getVersion` or `getFeatures`.

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (dbartol) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:82]: Fixed, although this was the same message we had before.

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (dbartol) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:140]: I think that should work. Fixed.

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (dbartol) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:154]: I suspect it was a typo. Fixed.

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (dbartol) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:171]: Easy enough to do now, since this is the only call site. See doc comment on `packBundle`

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (dbartol) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:294]: https://github.com/github/codeql-core/issues/4089

File: extensions/ql-vscode/test/vscode-tests/cli.ts
Comment (dbartol) [reply extensions/ql-vscode/test/vscode-tests/cli.ts:11]: I've made the change. I assume we're happy to assume these are file system paths?

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (dbartol) [reply extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts:34]: Done

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (dbartol) [reply extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts:40]: Done

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (dbartol) [reply extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts:44]: Done

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (dbartol) [reply extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts:45]: Done

**Comment by dbartol** [comment]
Thanks for all the reviews. I believe I've addressed all the feedback.

File: extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts
Comment (charisk) [reply extensions/ql-vscode/test/vscode-tests/cli-integration/variant-analysis/variant-analysis-manager.test.ts]: Thanks! Just a couple `\n` left on the message (L32).

File: extensions/ql-vscode/src/common/short-paths.ts
Comment (charisk) [reply extensions/ql-vscode/src/common/short-paths.ts:1]: I'll raise an issue to cover this

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (charisk) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts:294]: Nice one! Thanks for the link. Reminder that that's a private repo though 😬

File: extensions/ql-vscode/src/variant-analysis/run-remote-query.ts
Comment (charisk) [reply extensions/ql-vscode/src/variant-analysis/run-remote-query.ts]: Thank you, that's much more readable to me!

**Review by robertbrignull — APPROVED** [review]
This looks good to me, and I've given it another quick manual test.

Probably want to get an approval from @charisk too, but overall looks like this could be good to go and it was a very quick and nice turnaround.

**Review by charisk — APPROVED** [review]
Thanks for addressing the various comments and for getting this over the line! 🎉

**Pull request merged**
